compiler/AST/DecoratedClassType.cpp

```diff
@@ -88,28 +97,26 @@ DecoratedClassType* DecoratedClassType::copyInner(SymbolMap* map) {
  representation, which is used for most purposes within the compiler.
 
  */
-AggregateType* DecoratedClassType::getCanonicalClass() {
+AggregateType* DecoratedClassType::getCanonicalClass() const {
   INT_ASSERT(this->canonicalClass);
   return this->canonicalClass;
 }
 
 bool classesWithSameKind(Type* a, Type* b) {
   if (!a || !b) return false;
 
+  // TODO: handle managed pointers
   if (!isClassLike(a) || !isClassLike(b)) return false;
 
-  // AggregateType would mean borrow
-  ClassTypeDecorator aDecorator = CLASS_TYPE_BORROWED;
-  ClassTypeDecorator bDecorator = CLASS_TYPE_BORROWED;
-  if (DecoratedClassType* ad = toDecoratedClassType(a))
-    aDecorator = ad->getDecorator();
-  if (DecoratedClassType* bd = toDecoratedClassType(b))
-    bDecorator = bd->getDecorator();
+
+  // AggregateType would mean nonnil borrow
+  ClassTypeDecorator aDecorator = classTypeDecorator(a);
+  ClassTypeDecorator bDecorator = classTypeDecorator(b);
 
   return aDecorator == bDecorator;
 }
 
-Type* canonicalClassType(Type* t) {
+Type* canonicalDecoratedClassType(Type* t) {
```

Comment:
I understand canonical type, however how can a canonical type be decorated?

Please add a comment explaining what "canonical decorated" class type is and how this function relates to canonicalClassType(), getCanonicalClass(), getDecoratedClass().